src/ticket.c

```diff
@@ -314,8 +332,18 @@ int do_grant_ticket(struct ticket_config *tk, int options)
 
 	if (tk->leader == local)
 		return RLT_SUCCESS;
-	if (is_owned(tk))
-		return RLT_OVERGRANT;
+	if (is_owned(tk)) {
+		if (is_manual(tk) && (options & OPT_IMMEDIATE)) {
```

Comment:
Looks like a re-grant won't revoke the ticket first? If so, it means a revocation is expected to be requested by admin first, right? 

It'd be better if we could somehow figure out the current state of the ticket holder. If the current holder is still online, we probably shouldn't re-grant it anyway even if -F is supplied, i.e. enforced granting is allowed only if the holder is offline?

Reply:
The problem is that Booth generally doesn't know if the nodes are online or not.

I would like to add some idea of 'a site holding a ticket being actually online' by using heartbeats (which are sent for both automatic and manual tickets) but perhaps in a separate pull request, after agreeing with the upstream what exactly we want to have.

Perhaps then, we could leave this flow as it is now and open a new _something_ (issue?) to agree how we would like to track onlineness of Booth sites?

Reply:
Hmm, what do you mean "doesn't know"? There are heartbeats per ticket and booth establishes the state of all nodes on start and keeps tracking that state etc. Even though there won't be a failover, heartbeats should happen for manual tickets too in order for all parties to be able to display correct information about the ticket.

Reply:
Yes, you are right, heartbeats are still sent for manual tickets.
Currently, even if we don't receive a heartbeat for a long time, the node will not be marked as offline. It will loose the ticket but due to the fact that another node organizes and wins the elections for this particular ticket. I am not aware of any mechanism of marking a node as offline in Booth, but perhaps I am missing something?

Reply:
No, you're not. I see your point now. A node or, to be more precise, a site cannot be marked as offline. Every ticket has its own elections, heartbeats, and the rest. No tickets means no communication. It was inherited and I often considered changing it, but was always worried about introducing instability. Besides, it would be a very big modification. Probably something to look at for v2.0 :)

Reply:
Yeeah, exactly. Al least something like configuration parameter 'consider a site offline after not receiving heartbeats for XX time'.

Reply:
Well, booth could mark a member lost if it didn't hear from it in a while providing there are some tickets configured for which communication was expected. That shouldn't be too difficult to implement. Shouldn't ;-)

Reply:
Been thinking about this... Indeed another option is to consider the per-ticket state based on "timeout"/"expire" parameters. But since this is *manual* ticket,  it probably makes sense to keep the logic simple:

As long as there's any holder or potential holder of the ticket:
* Refuses to grant the ticket and gives a warning "Make sure the ticket has been revoked already!"
* With "-F", it grants the ticket with a warning like "the risk is on your own";-)
?

Reply:
Agreed.